client/boot/index.js

```diff
@@ -369,6 +370,30 @@ function reduxStoreReady( reduxStore ) {
 		require( 'lib/rubberband-scroll-disable' )( document.body );
 	}
 
+	// 404 handler.
+	page( '*', function( context, next ) {
+		let Page404 = require( 'layout/404.jsx' );
```

Comment:
The `.jsx` is not needed and it's kinda part of our convention to drop it (same for `.js`). So `layout/404` is enough :)